web/src/reducers/apd.js

```diff
@@ -77,6 +277,51 @@ const initialState = {
 
 const reducer = (state = initialState, action) => {
   switch (action.type) {
+    case ADD_APD_ITEM: {
+      const patches = getPatchesForAddingItem(state, action.path);
+      return {
+        ...state,
+        data: applyPatch(state.data, patches)
+      };
+    }
+
+    case ADD_APD_YEAR: {
+      const patches = getPatchesToAddYear(state, action.value);
+      return { ...state, data: applyPatch(state.data, patches) };
+    }
+
+    case EDIT_APD: {
+      return {
+        ...state,
+        data: applyPatch(state.data, [
+          {
+            op: 'replace',
+            path: action.path,
+            value: action.value
+          }
+        ])
+      };
+    }
+
+    case REMOVE_APD_ITEM: {
+      return {
+        ...state,
+        data: applyPatch(state.data, [
+          {
+            op: 'remove',
+            path: action.path
+          }
+        ])
+      };
+    }
+
+    case REMOVE_APD_YEAR: {
+      const patches = getPatchesToRemoveYear(state, action.value);
+      return { ...state, data: applyPatch(state.data, patches) };
+    }
+
+    // ------------------
```

Comment:
![I guess?](https://media3.giphy.com/media/K6VhXtbgCXqQU/giphy.gif)

Reply:
😂 
Above the line are reducers for the new actions, below the line are old reducers. I was (am?) operating on the assumption that a lot of the reducers below the line can be deleted.